processor/src/test/java/org/mapstruct/ap/test/generics/GenericsTest.java

```diff
@@ -8,27 +8,140 @@
 import org.mapstruct.ap.testutil.IssueKey;
 import org.mapstruct.ap.testutil.ProcessorTest;
 import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.compilation.annotation.CompilationResult;
+import org.mapstruct.ap.testutil.compilation.annotation.Diagnostic;
+import org.mapstruct.ap.testutil.compilation.annotation.ExpectedCompilationOutcome;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
 /**
  * @author Andreas Gudian
  *
  */
-@WithClasses({
-    AbstractTo.class,
-    AbstractIdHoldingTo.class,
-    Source.class,
-    SourceTargetMapper.class,
-    TargetTo.class
-})
 public class GenericsTest {
 
     @ProcessorTest
     @IssueKey("574")
+    @WithClasses({
+        AbstractTo.class,
+        AbstractIdHoldingTo.class,
+        Source.class,
+        SourceTargetMapper.class,
+        TargetTo.class
+    })
     public void mapsIdCorrectly() {
         TargetTo target = new TargetTo();
         target.setId( 41L );
         assertThat( SourceTargetMapper.INSTANCE.toSource( target ).getId() ).isEqualTo( 41L );
     }
+
+    @ProcessorTest
+    @IssueKey( "2690" )
+    @WithClasses({
+        TwinGenericTarget.class,
+        TwinGenericSource.class,
+        TwinGenericMapper.class
+    })
+    public void mapsIdenticalGenerics() {
+        TwinGenericSource<String, Integer> source = new TwinGenericSource<>();
+        source.setObject1( "StringObject" );
+        source.setObject2( 123 );
+
+        TwinGenericTarget<String, Integer> target = TwinGenericMapper.INSTANCE.map( source );
+
+        assertThat( target.getObject1() ).isEqualTo( "StringObject" );
+        assertThat( target.getObject2() ).isEqualTo( 123 );
+    }
+
+    @ProcessorTest
+    @IssueKey("2954")
+    @WithClasses({
+        GenericTargetContainer.class,
+        GenericSourceContainer.class,
+        GenericContainerMapper.class,
+    })
+    public void mapsSameTypeCorrectly() {
+        String containedObject = "Test";
+        GenericSourceContainer<String> container = new GenericSourceContainer<>( containedObject, "otherValue" );
+
+        GenericTargetContainer<String> resultString = GenericContainerMapper.INSTANCE.map( container );
+        assertThat( resultString.getContained() ).isSameAs( containedObject );
```

Comment:
Doing `isSameAs` with a String isn't really recommended. If we want to test this properly we should use some other object.